### PR TITLE
Add flatpak build

### PR DESCRIPTION
**Pull request by kdeleteme**
#788 

I wrote a manifest to build openspades with flatpak. This may help enable distribution through [flathub](https://flathub.org) if desired.

### CODE REVIEW

File: flatpak/jp.yvt.OpenSpades.appdata.xml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    first-person shooter created by Ben Aksoy, featuring fully destructible
+    terrain and plenty of game modes (including the well-known Capture the
+    Flag) created by the community.</p>
+    <ul>
+      <li>Can connect to a vanilla/pyspades/pysnip server</li>
+      <li>Uses OpenGL/AL for better experience</li>
+      <li>Open source, and cross platform</li>
+    </ul>
+  </description>
+
```
Comment (theunknownxy) [inline flatpak/jp.yvt.OpenSpades.appdata.xml]:
```suggestion

  <categories>
​      <category>Game</category>
​      <category>Shooter</category>
  ​</categories>

```

File: flatpak/jp.yvt.OpenSpades.appdata.xml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  <launchable type="desktop-id">jp.yvt.OpenSpades.desktop</launchable>
+
+  <url type="homepage">http://openspades.yvt.jp/</url>
+  <url type="bugtracker">https://github.com/yvt/openspades/issues</url>
+
```
Comment (theunknownxy) [inline flatpak/jp.yvt.OpenSpades.appdata.xml]:
```suggestion
  <url type="translate">https://crowdin.com/project/openspades</url>

```

File: flatpak/jp.yvt.OpenSpades.json [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+          "sha256": "c9e6723aad729140d4af0771d780d20823ca5b0c2fc52cf5a9b2028e923d33d4"
+        }
+      ]
+    },
+    {
+      "name": "openspades",
+      "buildsystem": "cmake",
+      "build-options": {
+        "env": {
+          "CC": "clang",
```
Comment (theunknownxy) [inline flatpak/jp.yvt.OpenSpades.json]: I almost forgot this: Is there are reason to use clang instead of gcc?

**Comment by feikname** [comment]
> For example i don't know whether "social-chat" also includes "language-profanity". The game itself doesn't contain any "bad" words, but the language on public servers usually isn't that nice.

I will check this later today.

File: flatpak/jp.yvt.OpenSpades.json (hunk 3 above)
Comment (kdeleteme) [reply flatpak/jp.yvt.OpenSpades.json]: I removed the build options.

File: flatpak/jp.yvt.OpenSpades.json [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  "cleanup": [
+    "/include",
+    "/lib/pkgconfig"
+  ],
+  "modules": [
+    "shared-modules/glew/glew.json",
+    {
+      "name": "opus",
+      "sources": [
+        {
```
[the remark this answers is not in the document]
Comment (kdeleteme) [reply flatpak/jp.yvt.OpenSpades.json]: I change it to pull the archive instead of from git.

**Comment by feikname** [comment]
From what I understood, `language-profanity` should only be used if the game itself contains profanity, and is not applicable to chat: that's why there's a separate `social-chat` attribute.

So the resulting OARS should be:
```xml
  <content_rating type="oars-1.1">
    <content_attribute id="violence-cartoon">intense</content_attribute>
    <content_attribute id="violence-bloodshed">mild</content_attribute>
    <content_attribute id="social-chat">intense</content_attribute>
  </content_rating>
```

**Comment by kdeleteme** [comment]
I edited the flatpak manifest to use the extra-data mechanism and added the OARS data. This looks good to me.

**Comment by feikname** [comment]
Is the issue where configuration/player data is reset on every launch still present?

**Comment by kdeleteme** [comment]
> Is the issue where configuration/player data is reset on every launch still present?

When I run it with flatpak-builder --run, it seems to save/load configuration data. But when I install and then run it, it resets the config.

**Comment by feikname** [comment]
>When I run it with flatpak-builder --run, it seems to save/load configuration data. But when I install and then run it, it resets the config.

Just a guess, but could you please try try changing `--persist=.local/share/openspades/` to `--persist=~/.local/share/openspades/`?

If that doesn't work either we can do `--filesystem=home` for now and do better sandboxing later (along solving #686).

**Comment by kdeleteme** [comment]
I tried your suggestions but they didn't work for me. I went with --filesystem=home

**Comment by feikname** [comment]
So, all good to merge? 😃

**Comment by kdeleteme** [comment]
Yup. Looks good to me. Finally :sweat_smile:

**Comment by theunknownxy** [comment]
Okay I have no idea what flatpak is doing: when you persist `.local/share/openspades`, it stores the files at `$HOME/.var/app/jp.yvt.OpenSpades/.local/share/openspades/Resources/SPConfig.cfg`. But within the container the openspades directory ended up at `$HOME/.local/share/openspades/openspades/`.
And suddenly after playing a bit around with the flatpak-builder config, the directory now ends up at the correct location...

Can you try to persist `.local/share/openspades` (without trailing slash) and then before you run OS delete `$HOME/.var/app/jp.yvt.OpenSpades/.local/`? Maybe flatpak was confused, because the directory already existed?

**Comment by kdeleteme** [comment]
Replaced `--filesystem=home` with `--persist=.local/share/openspades`

I tried it and it works for me.

**Comment by feikname** [comment]
@yvt this is ready to merge.

**Pull request merged**